Add new Modal Dialogue component

colinrotherham opened the pull request:
Afternoon 👋 

As discussed with @hannalaakso, @timpaul, @amyhupe our team has been working on contributing a new "Modal dialogue" component to the Design System.

Sadly I'm going to be handing over work on this to @FeyAgape and @Ash-Wilson so this is still a draft for now, but more work to come!

## Acceptance criteria

See https://github.com/alphagov/govuk-design-system-backlog/issues/30#issuecomment-553932293

## Changes to GOV.UK Frontend
To support a modal component I've had to enhance GOV.UK Frontend slightly:

* Adjusted iframe previews for `position: fixed` content
* Adjusted iframe previews for custom variant styling (fixed height for modals)
* Add support for Nunjucks macro call blocks `{% call macro() %}{% endcall %}`

But here's a draft pull request where we're up to anyway.

![Modal preview](https://user-images.githubusercontent.com/415517/69722029-09367680-110e-11ea-8d68-f81b8a2f6676.png)

colinrotherham commented:
Can we please have this on a Heroku preview URL so I can share with the team?

Thanks

colinrotherham commented:
Also any design feedback would be appreciated.

We're aware we added another embedded "crown" SVG, so you might want a better way of adding this to the page without duplicating the SVG markup.

I added the crown as it's already in use here:
https://components.publishing.service.gov.uk/component-guide/modal_dialogue

But not sure what the user need is.

soniaturcotte commented:
@alex-ju and I worked on one here, not sure if you've seen it: 
https://content-publisher.publishing.service.gov.uk/component-guide/modal_dialogue

colinrotherham commented:
> @alex-ju and I worked on one here, not sure if you've seen it:
> https://content-publisher.publishing.service.gov.uk/component-guide/modal_dialogue

Ahh is this the same code as used on as https://components.publishing.service.gov.uk/component-guide/modal_dialogue? If you're the original designer thank you 😁 

We've also improved the focus trapping so if you manage to click out of the modal, you're pulled back in again, not just looping from last/first items.

We've added `onOpen` and `onClose` callbacks, added native HTML5 `<dialog>` support, and turned it into a Nunjucks macro.

It's been great to improve one that's already in production.

nickcolley commented:
I have some questions on the design:

1. what is the user need for the crown logo on this modal?
2. we try to avoid using icons instead of words if possible as it is more accessible, has a word label been considered here?
3. is it clear that the close button is a button? should it have more visual affordance?

colinrotherham commented:
@nickcolley Maybe @soniaturcotte can help?

We used the design from https://components.publishing.service.gov.uk/component-guide/modal_dialogue and it tested well when did user research.

colinrotherham commented:
I've removed the vertical alignment fallback for older browsers using `display: table-cell` as it doesn't cater for overflowing content at all.

In IE10+ we've got basic flexbox support so that's good enough.

Older browsers like IE8 + IE9 will have the modal at the top of the screen.

colinrotherham commented:
@nickcolley Did you mean like this for the close button text?

![Close button text](https://user-images.githubusercontent.com/415517/69819127-25adde00-11f6-11ea-86c8-4e223e5b8e8f.png)

Maybe it's something @FeyAgape can explore?

You're in her capable hands now 😊

hannalaakso commented on src/govuk/components/modal-dialogue/modal-dialogue.yaml:
```diff
@@ -0,0 +1,120 @@
+          <p class="govuk-body">
+            This is some very long content.
+          </p>
+    call:
```
Could we name this `caller` to be consistent with [`caller` in the fieldset](https://github.com/alphagov/govuk-frontend/blob/master/src/govuk/components/fieldset/fieldset.yaml#L39)?

colinrotherham replied:
@hannalaakso Sounds good, I've changed it 👍

hannalaakso commented on app/app.js:
```diff
@@ -152,7 +152,7 @@ module.exports = (options) => {
     const requestedExampleName = req.params.example || 'default'
 
-    const previewLayout = res.locals.componentData.previewLayout || 'layout'
+    let previewLayout = res.locals.componentData.previewLayout
```
Non-blocking but as this variable is now used in the [iframe if-block](https://github.com/alphagov/govuk-frontend/blob/master/src/govuk/components/fieldset/fieldset.yaml#L39) we could consider moving it above that block.

colinrotherham replied:
@hannalaakso Yep sounds good too, all changed

hannalaakso commented on app/views/layouts/iframe.njk:
```diff
@@ -0,0 +1,21 @@
+{% block bodyEnd %}
+  {{ super() }}
+  {% block scripts %}{% endblock %}
+{% endblock %}
```
Would it make sense to call this file `component-iframe.njk`? We already have `component-preview.njk`.

colinrotherham replied:
@hannalaakso Done

colinrotherham commented on app/assets/scss/partials/_app.scss:
```diff
@@ -46,6 +53,7 @@
     content: " ";
     display: table;
     width: 100%;
+    height: 100%;
```
@hannalaakso Ah I thought that was a bug!

The blue outline now goes all the way around the iframe.

Because `iFrameResizer` can't calculate the height of a `position: fixed` modal, the fixed height comes from this bit that I've hard-coded:

```css
// Allow fixed height for "position: fixed" modal
.app-iframe-in-component-preview--modal {
  height: 400px;

  @include govuk-media-query ($from: tablet) {
    height: 500px;
  }
}
```

hannalaakso replied:
Going to resolve this as it doesn't concern the component itself. Good to have the conversation here for reference though.

colinrotherham commented on app/assets/scss/partials/_app.scss:
```diff
@@ -54,3 +62,12 @@
+
+  @include govuk-media-query ($from: tablet) {
+    height: 500px;
+  }
```
Thanks Hannah. If you could find a way to get `iFrameResizer` work for the modal that would be even better than a fixed height.

Looks like there's a `heightCalculationMethod` API method for iframe child content that could be customised? Currently it's not working due to `position: fixed`.

hannalaakso replied:
Going to resolve this as it doesn't concern the component itself. Good to have the conversation here for reference though.

colinrotherham commented on src/govuk/components/modal-dialogue/_modal-dialogue.scss:
```diff
@@ -0,0 +1,167 @@
+// Disables linting for this file only
+// sass-lint:disable no-duplicate-properties
```
@hannalaakso Annoyingly this feature was removed from `sass-lint`, but the documentation wasn't reverted. See https://github.com/sasstools/sass-lint/issues/1292

What looks like per-line disable rules actually affect the whole file! 🙈 

I'll move it anyway, as it's good to see which line it's trying to apply to, hopefully they enable it again in a future release.

colinrotherham commented on app/assets/scss/partials/_app.scss:
```diff
@@ -22,9 +22,16 @@
 
 .app-component-preview {
   position: relative;
-  width: 100%;
```
@hannalaakso Thanks I've put this back. It was the `margin: 5px` change that was causing the blue outline to appear on the left/right of full-width previews.

![After](https://user-images.githubusercontent.com/415517/69869913-24d48500-12a6-11ea-9a75-79ed8d897907.png)

nickcolley commented on src/govuk/components/modal-dialogue/modal-dialogue.js:
```diff
@@ -0,0 +1,186 @@
+  this.hasNativeDialog = 'showModal' in this.$dialogBox
+
+  // Allowed focussable elements
+  this.focussable = [
```
There is an 'inert' feature we could consider referring to when reviewing the focus trap functionality in this proposal:

https://github.com/WICG/inert

It ships in chromium, I'm not sure if it's a proper standard yet but worth thinking about.

colinrotherham replied:
Yeah sounds like a good enhancement.

You'd have to be _very_ careful and keep the modal outside `govuk-width-container` or whatever the page wrapper element was.

```html
<body>
    <div class="govuk-width-container" inert>…</div>
    <div data-module="govuk-modal-dialogue">…</div>
</body>
```

It could go horribly wrong if the modal was inside the `[inert]` wrapper.

hannalaakso replied:
The early prototype version of the modal that GOV.UK based theirs on [used both the `inert` attribute and `aria-hidden`](https://github.com/hannalaakso/accessible-timeout-warning/blob/0991356bd5ceba342e810e494e567fa61c1fe9b9/src/govuk/components/timeout-warning/timeout-warning.js#L228) to make the `<main>` wrapper non-actionable and this tested well with screen reader users. The modal markup was outside `<main>` and below `<body>`. 

Colin does have a valid point: we need to be clear in the guidance about where the modal markup is placed in the page content.

nickcolley commented on src/govuk/components/modal-dialogue/modal-dialogue.js:
```diff
@@ -0,0 +1,186 @@
+ModalDialogue.prototype.init = function (options) {
+  this.options = options || {}
+
+  this.open = this.handleOpen.bind(this)
```
I wonder if these method behinds should be in the constructor instead?

colinrotherham replied:
@nickcolley I try to defer anything that uses more CPU cycles until it's actually needed (like when `.init()` gets run) but it's probably negligible so don't mind.

hannalaakso replied:
Thanks Colin 👍 We believe that the performance impact is negligible so we should make the change for consistency with other components.

nickcolley commented on src/govuk/components/modal-dialogue/modal-dialogue.js:
```diff
@@ -0,0 +1,186 @@
+  this.$lastActiveElement.focus()
+
+  // Optional 'onClose' callback
+  if (typeof this.options.onClose === 'function') {
```
I wonder if we need to consider how we're documenting the JavaScript options for components, since we use data-attributes elsewhere instead of programmatic options.

See also: [Proposal programmatic and data attribute api interface for components](https://github.com/alphagov/govuk-frontend/pull/1238)

colinrotherham replied:
@nickcolley Yeah good point.

Here's how we currently use this Modal API on our prototype:

```js
var $dialogue = document.querySelector('[data-module="govuk-modal-dialogue"]')
var $dialogueButtonResume = document.querySelector('.govuk-modal-dialogue__continue')

if ($dialogue && $dialogueButtonResume) {
  var modal = new window.ModalDialogue($dialogue)
    .init({
      focusElement: $dialogueButtonResume,
      onClose: function () {
        // Extend session timeout here
      }
    })
}
```

hannalaakso commented on src/govuk/components/modal-dialogue/_modal-dialogue.scss:
```diff
@@ -0,0 +1,161 @@
+@import "../../helpers/all";
+
+@include govuk-exports("govuk/component/modal-dialogue") {
+  $govuk-dialogue-width: 640px;
```
We probably want to use [`!default`](https://sass-lang.com/documentation/variables#default-values) here so that the value can overridden.

hannalaakso commented on src/govuk/components/modal-dialogue/_modal-dialogue.scss:
```diff
@@ -0,0 +1,161 @@
+    box-sizing: border-box;
+    display: block;
+    position: relative;
+    z-index: 1;
```
We might want to increase this, just in the (non-ideal) case that the page has other competing elements. I notice that the [GOV.UK Publishing equivalent sets this to 1000](https://github.com/alphagov/govuk_publishing_components/blob/34315777fabc6ac53348e636c7d3006b62228fcd/app/assets/stylesheets/govuk_publishing_components/components/_modal-dialogue.scss#L3).

NogginBox replied:
I've made this 1001 so it is above the backdrop which also needs the z-index increasing.
```suggestion
    z-index: 1001;
```

hannalaakso commented on src/govuk/components/modal-dialogue/_modal-dialogue.scss:
```diff
@@ -0,0 +1,161 @@
+    width: 90%;
+    margin: auto;
+    padding: 0;
+    overflow-y: auto;
```
This might be okay as it is but just for reference, I notice that the GOV.UK Publishing example currently sets these to:
```
overflow-x: hidden;
overflow-y: scroll;
```
I don't think we currently have an example with page content in the main area so we need to add one in order to check that the main page content is not scrollable and the modal is.

hannalaakso commented on src/govuk/components/modal-dialogue/_modal-dialogue.scss:
```diff
@@ -0,0 +1,161 @@
+    background: govuk-colour("black");
+    pointer-events: none;
+    touch-action: none;
+  }
```
I think this block should live below https://github.com/alphagov/govuk-frontend/pull/1668/files#diff-40be96fa6566e06e9d9d9ab4d1d3d831R8 to in order to keep the backdrop styles together.

hannalaakso commented on src/govuk/components/modal-dialogue/_modal-dialogue.scss:
```diff
@@ -0,0 +1,161 @@
+    }
+  }
+
+  // New dialogue width, inline button + link
```
I'm not sure why these styles are set separately from the other ones in this file. Should this be a modifier class?

hannalaakso commented on src/govuk/components/modal-dialogue/_modal-dialogue.scss:
```diff
@@ -0,0 +1,161 @@
+  }
+
+  // Fixed width
+  @include govuk-media-query($from: desktop) {
```
I'm not sure why these styles are set separately from the other ones in this file. Should this be a modifier class?

hannalaakso commented on src/govuk/components/modal-dialogue/modal-dialogue.js:
```diff
@@ -0,0 +1,186 @@
+
+  // Close dialogue on close button click
+  this.$buttonClose.addEventListener('click', this.close)
+}
```
Not a big deal but I wonder if the events inside this block could live inside `init()`, it'll be tidier once we move the variable definitions to the constructor.

hannalaakso commented on src/govuk/components/modal-dialogue/modal-dialogue.js:
```diff
@@ -0,0 +1,186 @@
+  this.$lastActiveElement = document.activeElement
+
+  // Disable scrolling, show wrapper
+  this.$container.classList.add('govuk-!-scroll-disabled')
```
Does `govuk-!-scroll-disabled` do anything? I can't see any styles created for it.

hannalaakso commented on src/govuk/components/modal-dialogue/template.njk:
```diff
@@ -0,0 +1,69 @@
+          type: "button",
+          classes: "govuk-modal-dialogue__close",
+          attributes: {
+            "aria-label": "Close modal dialogue"
```
I wonder if all users will know what is meant by "modal dialogue"? The users of GOV.UK Publishing (which this is based on) would mainly be repeat users with at least a moderate level of computer literacy and are therefore not representative of the wider population.

If we could test with an alternative wording that would be great.

hannalaakso commented on src/govuk/components/modal-dialogue/template.njk:
```diff
@@ -0,0 +1,69 @@
+          viewbox="0 0 132 97"
+          height="30"
+          width="36"
+        >
```
Had a chat with @FeyAgape. We should consider removing the crown logo as there is no clear user need. Its usage here is not consistent with our other components and it adds complexity. 

Its usage in public facing sites in this manner might also be problematic with regards to the branding guidelines; the [Service Manual guidance](https://www.gov.uk/service-manual/design/making-your-service-look-like-govuk#if-your-service-isnt-on-govuk) only mentions the use of the crown logo in the header.

hannalaakso commented on src/govuk/components/modal-dialogue/template.njk:
```diff
@@ -0,0 +1,69 @@
+        </svg>
+
+        {{ govukButton({
+          text: "×",
```
Had a chat with @FeyAgape. If possible, we should explore using a text equivalent here as icons might not be understood by all users.

hannalaakso commented on src/govuk/components/modal-dialogue/modal-dialogue.yaml:
```diff
@@ -0,0 +1,120 @@
+previewLayout: modal
+
+examples:
+  - name: default
```
Could we make this example to open the modal with a button? Similar to the [GOV.UK Publishing modal example](https://components.publishing.service.gov.uk/component-guide/modal_dialogue/preview). This would allow us to test the transition of the modal being closed to it being open, and how the main page content is made actionable / non-actionable.

Doing this might require some tweaking of the review app.

hannalaakso commented on src/govuk/components/modal-dialogue/modal-dialogue.yaml:
```diff
@@ -0,0 +1,120 @@
+      description:
+        html: <p class="govuk-body">Prompt text for the modal dialogue component goes here</p>
+
+  - name: open with scrolling content
```
Could we make this example to have some page content in the main page area?  Similar to the [GOV.UK Publishing modal example](https://components.publishing.service.gov.uk/component-guide/modal_dialogue/preview). This would allow us to test that the background is not scrollable / actionable when the modal is open.

hannalaakso commented on src/govuk/components/modal-dialogue/modal-dialogue.yaml:
```diff
@@ -0,0 +1,120 @@
+- name: open
+  type: boolean
+  required: true
+  description: If true, modal dialogue element will be visible.
```
Should we consider calling the component "modal dialog" instead? Most of the industry docs talk about "modal dialog" or "dialog element":
https://www.w3.org/TR/wai-aria-practices/examples/dialog-modal/dialog.html
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dialog

nickcolley replied:
https://trends.google.com/trends/explore?cat=422&q=modal,dialog,modal%20dialog,modal%20dialogue,dialogue

nickcolley commented on src/govuk/components/modal-dialogue/modal-dialogue.js:
```diff
@@ -0,0 +1,186 @@
+}
+
+// Ensure focus stays within modal
+ModalDialogue.prototype.handleFocusTrap = function (event) {
```
To add to this point, what if a user navigates via landmarks? What is the impact on them when the modal is open.

nickcolley commented:
@hannalaakso I can help review this when you're happy later on. I think we should do some real device testing at some point too.

NogginBox commented:
What is the status of this PR?

We're adding a timeout modal in our next sprint and I would like to use this component.

hannalaakso commented:
@NogginBox Thanks for reaching out. This work depends on https://github.com/alphagov/govuk-frontend/issues/1708 especially, and also https://github.com/alphagov/govuk-frontend/issues/1722. It's unfortunately not possible to give a timeline on the work at the moment but we will pick it up when we're able to prioritise it.

NogginBox commented:
I've used the code from this PR to get a modal dialogue timeout warning in our site.

I've made one change to it, which which is to increase the z-index to 100 to make sure that it is above all other elements of the site.

NogginBox commented on src/govuk/components/modal-dialogue/_modal-dialogue.scss:
```diff
@@ -0,0 +1,161 @@
+  .govuk-modal-dialogue,
+  .govuk-modal-dialogue__backdrop {
+    position: fixed;
+    z-index: 0;
```
Increase to 1000 so that this will site above all other elements already in the site.

I had to do this when I've used this code as several elements floated on top of the backdrop.

NogginBox commented on src/govuk/components/modal-dialogue/_modal-dialogue.scss:
```diff
@@ -0,0 +1,161 @@
+
+  // Inner content
+  .govuk-modal-dialogue__content {
+    @include govuk-font($size: 16);
```
I was able to safely remove this font definition as I wanted the site's standard fonts to be used. We're a non gov body so can't use the standard GDS font and have overrides in place to not use them. This line being included broke our GDS font overrides.
```suggestion
```

zaizous commented:
Hi @hannalaakso, do you guys have any update on the timeline? 
We are looking into implementing time out in the coming sprints and would be good to make use of the GDS pattern.

hannalaakso commented:
@zaizous No update at the moment. You might find the [acceptance criteria which this component will need to meet](https://github.com/alphagov/govuk-design-system-backlog/issues/30#issuecomment-553932293) helpful if you're researching this topic.

hannalaakso commented:
I’m closing this PR as it’s currently blocked by #1722 and #1708.

Note for teams who are researching modal dialogues: the work done in this PR could be used as the basis for creating a modal dialogue. We would encourage teams to review: 
- [acceptance criteria related to this PR](https://github.com/alphagov/govuk-design-system-backlog/issues/30#issuecomment-553932293), especially the points about design and accessibility
- conversations on the [modal dialogue backlog issue
](https://github.com/alphagov/govuk-design-system-backlog/issues/30)
